dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/data/batching/DefaultConsentAwareDataHandler.kt

```diff
@@ -0,0 +1,49 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.data.batching
+
+import com.datadog.android.core.internal.data.privacy.Consent
+import com.datadog.android.core.internal.data.privacy.ConsentProvider
+import com.datadog.android.core.internal.data.privacy.ConsentProviderCallback
+
+internal class DefaultConsentAwareDataHandler<T>(
+
+    private val consentProvider: ConsentProvider,
+    private val intermediaryDataProcessor: DataProcessor<T>,
+    private val targetDataProcessor: DataProcessor<T>,
+    private val batchedDataMigrator: BatchedDataMigrator
+
+) : ConsentAwareDataHandler<T>, ConsentProviderCallback {
+
+    init {
+        consentProvider.registerCallback(this)
+    }
+
+    @Synchronized
+    override fun consume(event: T) {
+        when (consentProvider.getLastConsent()) {
+            Consent.PENDING -> {
+                intermediaryDataProcessor.consume(event)
+            }
+            Consent.GRANTED -> {
+                targetDataProcessor.consume(event)
+            }
+            else -> {
+                // do nothing
+            }
+        }
+    }
+
+    @Synchronized
+    override fun onConsentUpdated(lastConsent: Consent, newConsent: Consent) {
+        if (newConsent == Consent.GRANTED && lastConsent == Consent.PENDING) {
+            batchedDataMigrator.moveData()
+        } else if (newConsent == Consent.NOT_GRANTED) {
+            batchedDataMigrator.clearData()
+        }
+    }
```

Comment:
These two are synchronized individually, but we should also synchronize their exclusive operations, no? Otherwise, data may be moved while being written.

Reply:
I see your point and it's actually something that I need to consider here. I could actually put everything in a WorkerThread here and don't synchronise it anymore as all the ops will perform in serial. Then all the ops inside the DataProcessor and DataMigrator will be synchronised.
This will work and keep the things simple but will mean that the DataProcessor and DataMigrator will not use any WorkerThread and will be only called from the ConsentAwareDataHandler.

On the other hand if I think this through say we are moving data now and then we get a new event. The DataMigrator is still working to move the data to the upload folder but in the same time the DataProcessor (the target one in this case because the flag was GRANTED already) wants to write in that folder. I don't see a problem here as the DataProcessor will definitely write in a different file than the ones we are trying to move and also the  system already the file for access while the InputStream is opened. 
The main reason why I wanted to have the synchronisation here is to make sure that after the data started to be moved we are using the right DataProcessor to write any event after otherwise we are going to loose some (they will be written with the old DP and there's not going to be any data migration after).

Reply:
For your last point, you don't need that much syncronisation, the consent flag is an atomic value, meaning as soon as you change it the data will be sent to the right processor. You can event schedule the migration, meaning that you'll never miss any data